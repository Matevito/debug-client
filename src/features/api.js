const axios = require('axios');

const api = axios.create({
    baseURL: "https://pure-falls-26749.herokuapp.com/apiv1/",
    timeout: 3000
});

export default api;