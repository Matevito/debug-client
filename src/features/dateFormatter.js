import { DateTime } from "luxon"

const simpleDate = (string) => {
    return DateTime.fromISO(string).toLocaleString(DateTime.DATE_MED)
};

const detailedDate = (string) => {
    return DateTime.fromISO(string).toLocaleString(DateTime.DATETIME_FULL_WITH_SECONDS)
}

export {
    simpleDate,
    detailedDate
}