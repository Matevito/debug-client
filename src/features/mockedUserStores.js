import configureStore from "redux-mock-store";

// todo: set project an issues values
const usersStoreValues = [
    {
        user: {} 
    },
    {
        user: {
            loggedIn: true,
            token : "a token?",
            user: {
                username: "devUser",
                email:"dev@email.com",
                role: "Developer",
                id: "1"
            },
            projects: {
                number: 0,
                list: []
            },
            issues: {
                number: 0,
                list: []
            }
        }
    },
    {

        user: {
            loggedIn: true,
            token : "a token?",
            user: {
                username: "TeamLeaderUser",
                email:"team@email.com",
                role: "Team leader",
                id: "2"
            },
            projects: {
                number: 0,
                list: []
            },
            issues: {
                number: 0,
                list: []
            }
        }
    },
    {

        user: {
            loggedIn: true,
            token : "admintoken?",
            user: {
                username: "adminUser",
                email:"admin@email.com",
                role: "Admin",
                id: "3"
            },
            projects: {
                number: 0,
                list: []
            },
            issues: {
                number: 0,
                list: []
            }
        }
    }
];

exports.getData = () => {
    const mockedUsers = usersStoreValues.map((user, i) => {
        const mockStore = configureStore([]);
        if (i === 0) {
            return mockStore(user)
        }
        return mockStore({user: {
            user: {
                user
            }
        }})
    });

    return mockedUsers;
}