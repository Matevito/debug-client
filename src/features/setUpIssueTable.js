const setUpIssueTable = (issuesList) => {
    let data = [
        {label: "open", x: 1,  y: 0},
        {label: "aditional info needed", x: 2, y: 0},
        {label: "in progress", x: 3, y: 0},
        {label: "under review", x: 4, y: 0},
        {label: "solved",x: 5, y: 0}
    ];

    issuesList.forEach((issue) => {
        data.forEach((obj, index) => {
            if (obj.label === issue.status){
                data[index].y += 1
            }
        })
    });
    
    return data.filter(obj => obj.y > 0);
};


export default setUpIssueTable;