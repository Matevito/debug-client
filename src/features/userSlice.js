import  { createSlice } from "@reduxjs/toolkit";

/* User object structure
user:{
    loggedIn: 'boolean',
    user: {
        username,
        email,
        role,
        id
    },
    projects: [{
        id,
        title
    }],
    issues: [{
        id,
        title
    }],
    todo on it's own slice:?
    notifications: [array of id's] // under development*
}
*/
export const userSlice = createSlice({
    name: 'user',
    initialState: {
        user: null
    },
    reducers: {
        login: (state, action) => {
            state.user = action.payload
        },
        logout: (state) => {
            state.user = null
        }
    }
});

export const { login, logout } = userSlice.actions;

export const selectUser = (state) => state.user.user;

export default userSlice.reducer