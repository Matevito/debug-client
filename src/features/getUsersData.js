import api from "./api";

const getUserData = async (token) => {
    const config = {
        headers: { "auth-token" : token}
    }; 
    try {
        const apiRes = await api.get("/user/list", config);
        return apiRes.data.data
    } catch(err) {
        return null
    }
}

export default getUserData