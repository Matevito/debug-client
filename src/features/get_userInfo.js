import api from "./api";

const get_userInfo = async (token) => {
    // setting up headers for calls
    const config = {
        headers: {
            "auth-token": token
        }
    };
    // fetch api data
    const userData = await api.get("/whoami", config);
    if (!userData) {
        return false;
    }
    const userId = userData.data.data.id;
    const userRelatedData = await api.get(`/user/${userId}`, config);
    if (!userRelatedData) {
        return false;
    };

    const userNotifications = [];

    // structure res and send it
    const dataRes = {
        loggedIn: true,
        token: token,
        user: {
            username: userData.data.data.username,
            email: userData.data.data.email,
            role: userData.data.data.role,
            id: userData.data.data.id
        },
        projects: userRelatedData.data.data.projects,
        issues: userRelatedData.data.data.issues,
        notifications: {
            list: userNotifications,
            number: userNotifications.length
        }
    };
    return dataRes;
}

export default get_userInfo;