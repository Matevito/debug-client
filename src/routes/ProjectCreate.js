import React, { useEffect, useState } from 'react'
import { useNavigate } from "react-router-dom";

// redux state management
import { useSelector } from "react-redux";
import { selectUser } from "../features/userSlice";
import { useDispatch } from "react-redux";
import { login } from "../features/userSlice";

// mui components
import {
    Box,
    Typography
} from "@mui/material"
import AssignmentIcon from '@mui/icons-material/Assignment';

// app components
import { ProjectForm } from "../components/ProjectForm";

// api comp
import api from "../features/api";
import get_userInfo from "../features/get_userInfo";

export const ProjectCreate = () => {
    const user = useSelector(selectUser);
    let navigate = useNavigate();
    const dispatch = useDispatch();

    const [usersList, setUsersList] = useState([]);
    const [errors, setErrors] = useState(null)

    useEffect(() => {
        async function getUsersList(){
            if(user.user.role === "Admin") {
                const config = {
                    headers: {
                        "auth-token": user.token
                    }
                };
                
                const apiRes = await api.get("/user/list", config);
                if (apiRes.status === 200) {
                    setUsersList(apiRes.data.data)
                }
            }
        };
        if (user) {
            getUsersList()
        }
    },[user]);

    const handleSubmit = async (form) => {
        /* if it fails set errors, if succed navigate to '/' */
        const url = '/project/';
        const config = {
            headers: {
                "auth-token": user.token
            }
        }
        console.log(config)
        try {
            await api.post(url, form, config);
            
            // if the req succeds, refresh user data and navigate to '/'
            const userData = await get_userInfo(user.token);
            dispatch(login(userData));
            navigate("/");
        } catch (err) {
            setErrors(err.response.data.error)
            console.log(err.response.data)
        }
    }
    if (!user) {
        // nvigate to /home
        navigate("/")
    } else if (user.user.role !== "Admin") {
        //navigate to "/protected-route"
        navigate("/protected-route")
    } else {
        return (
            <Box
                sx={{
                    marginTop: 2,
                    display: "flex",
                    flexDirection: "column",
                    alignItems: 'center',
                }}
                fullWidth
            >
                <AssignmentIcon />
                <Typography component="h1" variant="h5">
                    Create Project
                </Typography>
                <p></p>
                <ProjectForm  
                    usersList={usersList}
                    errors={errors}
                    handleSubmit={handleSubmit}
                />
            </Box>
        )
}
}
