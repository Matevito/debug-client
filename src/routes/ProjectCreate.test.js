import React from "react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";

//testing com
import { render, screen } from "@testing-library/react";
import '@testing-library/jest-dom'

// test comp
import { ProjectCreate } from "./ProjectCreate";

// mock functs
import { getData } from "../features/mockedUserStores";

const mockedUsedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUsedNavigate,
}));

describe("ProjectCreate component", () => {
    let noLoggedUser;
    let devStore;
    let teamLstore;
    let adminStore;

    beforeEach(() => {
        const usersData = getData();

        noLoggedUser = usersData[0];
        devStore = usersData[1];
        teamLstore = usersData[2];
        adminStore = usersData[3];
    })
    test("handles no logged user", () => {
        const { container } = render(<Provider store={noLoggedUser}><ProjectCreate /></Provider>, {wrapper: MemoryRouter});
        // does no renders anything;
        expect(container).toMatchSnapshot();
        expect(mockedUsedNavigate).toHaveBeenCalledWith("/");
    })
    test("handles developer user", () => {
        // developer user
        const { devContainer } = render(<Provider store={devStore}><ProjectCreate /></Provider>, {wrapper: MemoryRouter});
        expect(devContainer).toMatchSnapshot();
        expect(mockedUsedNavigate).toHaveBeenCalledWith("/protected-route");
    });
    test("handles team leader user", () => {
        const { teamContainer } = render(<Provider store={teamLstore}><ProjectCreate /></Provider>, {wrapper: MemoryRouter});
        expect(teamContainer).toMatchSnapshot();
        expect(mockedUsedNavigate).toHaveBeenCalledWith("/protected-route");
    })
    test.todo("renders complete form");
}) 