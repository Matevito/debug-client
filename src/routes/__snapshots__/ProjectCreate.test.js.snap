// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectCreate component handles developer user 1`] = `undefined`;

exports[`ProjectCreate component handles no admin user 1`] = `undefined`;

exports[`ProjectCreate component handles no logged user 1`] = `<div />`;

exports[`ProjectCreate component handles team leader user 1`] = `undefined`;
